library/alloc/src/rc.rs

```diff
@@ -2208,13 +2208,11 @@ impl<T: ?Sized> Unpin for Rc<T> {}
 /// This has the same safety requirements as `align_of_val_raw`. In effect:
 ///
 /// - This function is safe for any argument if `T` is sized, and
-/// - if `T` is unsized, the pointer must have appropriate pointer metadata
-///   acquired from the real instance that you are getting this offset for.
+/// - if `T` is unsized, the pointer must have been acquired from an unsize
+///   coercion (but may be invalid, such as from Weak::new).
```

Comment:
This is not "the same safety requirements as `align_of_val_raw`", though. `align_of_val_raw` quite deliberately only talks slices, trait objects, and `extern` types -- for all other unsiezd tails, "it is not allowed to call this function".

Reply:
This is super awkward to specify cleanly without copying over the entire docs from `align_of_val_raw`. When I made this adjustment, I figured that "comes from an unsize coercion" would be a tighter bound than `align_of_val_raw` spells out, since the only types of unsize coercion are for slices and trait objects. "Comes from unsize coercion" is in fact a simpler spelling of the requirement for trait objects (valid vptr). For slices, it's exactly equivalent iff it's impossible to name a fixed-size array type that has a layout close to the size of `isize::MAX` (such that `RcInner<T>` is too big). I think that the error "values of the type `rc::RcBox<[u128; N]>` are too big for the current architecture" is enough for that, though I do think relying on said error for soundness is suboptimal.

(Also, we could potentially relax `align_of_val_raw` to limit the magnitude of alignment (which it probably should anyway) rather than size.)

Reply:
The reason I don't like "comes from an unsize coercion" is that it is open-ended -- it is true for the unsizing coercions that currently exist, but that  set might change in the future. The `align_of_val_raw` docs are quite deliberately stated in a way that is *not* open-ended in this way.